# Documentation - standardScaler

## Introduction
Ce notebook présente la documentation complète du StandardScaler, 
une technique de standardisation des données utilisée dans les 
pipelines de Machine Learning. Vous y trouverez la description, 
la formule, le fonctionnement de chaque méthode, des exemples 
de code et une comparaison avec Scikit-learn.

## Description
Le standardScaler est utilisé pour standariser les données c'est a dire les  mettre a la meme échelle. Il calcule la moyenne et l"écart-type de chaque colonne puis transforme les valeurs pour que chaque colonne ait une moyenne de 0 et un écart-type de 1.

### Formule
***X' = (X - moyenne) / écart-type.***

La mise en place et le fonctionnement du standardScaler se base de six (06) méthode dont: fit(); transform(); fit_transform(); inverse_transform(); convert_to_array(); _convert_to_dataframe().

## Méthode

### fit()
La méthode fit() a pour role de calculer et de mémoriser la moyenne et l'écart-type de chaque colonne. Elle prend en entrée un tableau de données X, calcul la moyenne avec np.mean() et l'écart-type avec np.std() pour chaque colonne puis il les mémorise dans self.mean() et self.std()afin de retourner le scaler.

###  transform ()
La méthode transform() a pour role de standariser les données en appliquant la formule Z-score. Elle prend en entrée un tableauX, vérifie que fit() a bien été appelé, applique la formule X' = (X - moyenne) / écart-type et retourne les données standarisées dans le meme format que l'entrée.

### fit_transform()
La méthode fit_transform est une union des deux première méthode.Elle appelle fit(X) pour calculer la moyenne et l'écart-type, puis transform(X) standarise et retourne les données standarisées.

### inverse_transform()
La méthode inverse_transform() a pour role de remettre les données dans leurs échelle d'origine. Elle prend des données standarisées, Vérifie que fit() a été appelé, applique la formule inverse (X* écart-type) + moyenne et retourne les données originales.

### convert_to_array()
La méthode convert_to_array() a pour role de convertir les données d'entrée en tableau numpy. Elle prend en entrée des données X et les convertit dans le but de permettre les calculs. Si X est un dataframe, elle mémorise les noms de colonne afin de retourner les valeurs. Dans d'autres cas, elle convertit directement en tableau numpy.

### convert_to_dataframe()
La méthode convert_to_dataframe() a pour role de remettre les données dans leurs format d'origine. Elle prend en entrée le tableau numpy transformé afin de le reconvertit vers le format initial. Dans le cas ou les données originales était un dataframe, elle mémorise les noms de colonne et recrée un dataframe. Sinon, elle retourne le tableau numpay sans rien changé.





 Voici un exemple concret de l'utilisation du standardScaler et une comparaison faites avec scikit-learn:

In [1]:
import sys
sys.path.append(r'C:\Users\LENOVO\ifri_mini_ml_lib\ifri_mini_ml_lib\preprocessing\preparation\scaler')

from standard_scaler import StandardScaler

print("Import réussi !")

Import réussi !


In [4]:

import numpy as np
import pandas as pd
import time
from sklearn.preprocessing import StandardScaler as SklearnStandardScaler

### Chargement du dataset
df = pd.read_csv('base3.csv')

### Sélection de 3 colonnes numériques
data = df[['Age', 'EstimatedSalary', 'Purchased']]

print("Dataset original :")
print(data.head(10))

### Notre implémentation
debut = time.time()
notre_scaler = StandardScaler()
notre_resultat = notre_scaler.fit_transform(data)
notre_temps = time.time() - debut

### Scikit-learn
debut = time.time()
sklearn_scaler = SklearnStandardScaler()
sklearn_resultat = sklearn_scaler.fit_transform(data)
sklearn_temps = time.time() - debut

print("Notre StandardScaler :")
print(notre_resultat)
print(f"Temps : {notre_temps:.6f} secondes")

print("\nScikit-learn StandardScaler :")
print(sklearn_resultat)
print(f"Temps : {sklearn_temps:.6f} secondes")

print("\nRésultats identiques :", np.allclose(notre_resultat, sklearn_resultat))

Dataset original :
   Age  EstimatedSalary  Purchased
0   19            19000          0
1   35            20000          0
2   26            43000          0
3   27            57000          0
4   19            76000          0
5   27            58000          0
6   27            84000          0
7   32           150000          1
8   25            33000          0
9   35            65000          0
Notre StandardScaler :
          Age  EstimatedSalary  Purchased
0   -1.781797        -1.490046  -0.745936
1   -0.253587        -1.460681  -0.745936
2   -1.113206        -0.785290  -0.745936
3   -1.017692        -0.374182  -0.745936
4   -1.781797         0.183751  -0.745936
..        ...              ...        ...
395  0.797057        -0.844019   1.340598
396  1.274623        -1.372587   1.340598
397  1.179110        -1.460681   1.340598
398 -0.158074        -1.078938  -0.745936
399  1.083596        -0.990844   1.340598

[400 rows x 3 columns]
Temps : 0.001819 secondes

Scikit-learn Stand

## Analyse des résultats

### Résultats de la transformation
Les deux scalers produisent des résultats identiques (True).
Les données sont bien centrées autour de 0 :
- Les valeurs d'âge varient entre -1.78 et +1.27
- Les salaires varient entre -1.49 et +1.34
- La colonne Purchased donne -0.75 (non acheté) et +1.34 (acheté)

### Performances temporelles
- Notre StandardScaler : 0.002559 secondes
- Scikit-learn : 0.012293 secondes

Notre implémentation est environ 5 fois plus rapide que 
Scikit-learn sur ce dataset de 400 lignes.

## Noté Bien

### MinMaxScaler
- Ramène les valeurs entre 0 et 1
- Utilise la formule (X - min) / (max - min)
- Très sensible aux valeurs extrêmes
- Garde des bornes fixes [0, 1]
- À utiliser quand on connaît les bornes des données

### StandardScaler
- Centre les valeurs autour de 0
- Utilise la formule (X - moyenne) / écart-type
- Moins sensible aux valeurs extrêmes
- Pas de bornes fixes
- À utiliser quand les données suivent une distribution normale


## Conclusion
Notre StandardScaler est correct et plus rapide que Scikit-learn 
sur de petits jeux de données.